File: leetcode/Recover Binary Search Tree.cpp
// copied from leetcode, undone

/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void recoverTree(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
		TreeNode *f1 = NULL, *f2 = NULL, *pre = NULL, *par = NULL;
		bool found = false;
		while (root)
		{
			if (NULL == h->left)
			{
				if (par && par->val > h->val)
				{
					if (! found)
					{
						f1 = par;
						found = true;
					}
					f2 = root;
				}
				par = root;
				root = root->right;
				continue;
			}
		}
    }
};

File: answers/leetcode/Recover Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def recoverTree(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """

        wrong_nodes = [None, None]
        self.findWrongNodes(root, None, wrong_nodes)
        wrong_nodes[0].val, wrong_nodes[1].val = wrong_nodes[1].val, wrong_nodes[0].val

    def findWrongNodes(self, root, prev, wrong_nodes):
        if root:
            prev1 = self.findWrongNodes(root.left, prev, wrong_nodes)
            if prev1:
                prev = prev1
            if prev and prev.val > root.val:
                if None == wrong_nodes[0]:
                    wrong_nodes[0] = prev
                wrong_nodes[1] = root
            next = self.findWrongNodes(root.right, root, wrong_nodes)
            return next if next else root
        return root
